### PR TITLE
Nginx grpc errorcodes

### PR DESCRIPTION
**Pull request by JonathanWamsley**
When the receiver encounters an error, include a grpc status code. This can be used to generate a user facing error message.

HTTP status codes are interpreted according to the [official HTTP to gRPC mapping](https://grpc.github.io/grpc/core/md_doc_http-grpc-status-mapping.html). Other errors are interpreted according to [gRPC guidelines](https://grpc.github.io/grpc/core/md_doc_statuscodes.html).

`httptest` is used to trigger different responses, and zap/zaptest/observer was used to validate that appropriate log messages are generated.

### CODE REVIEW

**Review by djaglowski — APPROVED** [review]

File: go.mod [hunk 1]
```diff
@@ -82,22 +82,22 @@ require (
...
+	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/signalfxreceiver v0.0.0-00010101000000-000000000000
+	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/simpleprometheusreceiver v0.0.0-00010101000000-000000000000
+	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/splunkhecreceiver v0.0.0-00010101000000-000000000000
+	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/statsdreceiver v0.0.0-00010101000000-000000000000
+	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/syslogreceiver v0.0.0-00010101000000-000000000000
+	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/tcplogreceiver v0.0.0-00010101000000-000000000000
+	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/udplogreceiver v0.0.0-00010101000000-000000000000
+	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/wavefrontreceiver v0.0.0-00010101000000-000000000000
+	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/windowsperfcountersreceiver v0.0.0-00010101000000-000000000000
```
Comment (bogdandrutu) [inline go.mod:98]: Why this change? We have replace statements so the version does not matter.

File: receiver/nginxreceiver/scraper.go [hunk 2]
```diff
@@ -16,13 +16,16 @@ package nginxreceiver
...
+	"fmt"
 	"net/http"
+	"strings"
 	"time"
 
 	"github.com/nginxinc/nginx-prometheus-exporter/client"
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/model/pdata"
 	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
```
Comment (bogdandrutu) [inline receiver/nginxreceiver/scraper.go:28]: Why using gRPC codes?

**Review by bogdandrutu — COMMENTED** [review]
This is a "scraper" receiver, so this is not returned to the caller, and we log already the error and we have scrapper errors defined.

Hard to understand the reasoning to use gRPC errors in this case.

File: receiver/nginxreceiver/scraper.go (hunk 2 above)
Comment (jsuereth) [reply receiver/nginxreceiver/scraper.go:28]: I think the idea here is to use a consistent non-string error code to denote common failures and diagnostics across receivers.

I can write up a proposal around this if you'd like.

Comment (bogdandrutu) [reply receiver/nginxreceiver/scraper.go:28]: Please do, but in Golang this is usually what "error" does. We can define custom errors as we already did for "scrapererror" :) No need to use "grpc" codes here I think.

**Comment by tigrannajaryan** [comment]
@JonathanWamsley  please address @bogdandrutu 's comment.

File: go.mod (hunk 1 above)
Comment (JonathanWamsley) [reply go.mod]: Updated version

File: receiver/nginxreceiver/scraper.go (hunk 2 above)
Comment (jsuereth) [reply receiver/nginxreceiver/scraper.go:28]: Sorry for long delay here.   Yes, I think doing what was done for scrapererror is exactly what would help.   Just proposed gRPC codes for expediency since they're used in OTLP.

Comment (jsuereth) [reply receiver/nginxreceiver/scraper.go:28]: Looking at details, I think we're missing something in scrapererror, wrote more down here: https://github.com/open-telemetry/opentelemetry-collector/issues/4156
